File: 백준/Silver/2644. 촌수계산/촌수계산.kt
package org.example

import java.util.StringTokenizer

fun main() {
    val br = System.`in`.bufferedReader()
    val n = br.readLine().toInt() // 전체 사람 수 n
    val (p1, p2) = br.readLine().split(" ").map { it.toInt() }
    val m = br.readLine().toInt() // 관계 수 m
    val graph = Array(n + 1) { mutableListOf<Int>() }
    repeat(m) {
        val st = StringTokenizer(br.readLine())
        val a = st.nextToken().toInt()
        val b = st.nextToken().toInt()
        graph[a].add(b)
        graph[b].add(a)
    }
    val visited = BooleanArray(n + 1)
    println(dfs(graph, p1, p2, visited, 0))

    br.close()
}

fun dfs(graph: Array<MutableList<Int>>, v: Int, p2: Int, visited: BooleanArray, depth: Int): Int {
    visited[v] = true
    //println("v $v / depth $depth / p2 $p2")
    if (v == p2) {
        return depth
    }
    for (neighbor in graph[v]) {
        if (!visited[neighbor]) {
            val ret = dfs(graph, neighbor, p2, visited, depth + 1)
            if (ret != -1) {
                return ret
            }
        }
    }
    return -1
}

File: 백준/Silver/2644. 촌수계산/촌수계산.py
import sys

sys.setrecursionlimit(10 ** 6)
input = sys.stdin.readline
isDone = False


def dfs(graph, p1, p2, visited, count):
    ret = 0
    visited[p1] = True
    global isDone
    if p1 == p2:
        isDone = True
        return count

    for neighbor in graph[p1]:
        if not visited[neighbor]:
            if not isDone:
                ret = dfs(graph, neighbor, p2, visited, count + 1)
            else:
                return ret
    return ret


n = int(input())
p1, p2 = map(int, input().split())
m = int(input())
graph = [[] for _ in range(n + 1)]
visited = [False] * (n + 1)
for _ in range(m):
    a, b = map(int, input().split())
    graph[a].append(b)
    graph[b].append(a)

count = 0
ret = dfs(graph, p1, p2, visited, count)

print(-1 if ret == 0 else ret)
